### PR TITLE
core: move interop to a separate package

### PR DESCRIPTION
**Pull request by fyrchik**
This makes it able to define interops in packages which do not depend from `core`
E.g. interops from #777 and #779 can be defined in separate packages.

### CODE REVIEW

**Comment by fyrchik** [comment]
Last commit with `enumerator`/`iterator` interops is for illustrative purposes.

**Comment by roman-khimov** [comment]
I'm thinking now about `dao` and `Blockchain` responsibilities, technically interop context (which is almost an `AppExecContext`) can work without `Blockchain` for almost all cases except `BlockHeight/CurrentBlockHash` (passing `topBlock` to the context solves it easily) and `GetHeaderHash`, but even that can be reworked to go through `dao`. But what's the purpose of `Blockchain` then?

**Comment by fyrchik** [comment]
This is probably right, because everything which is not in top block, have already been pushed to storage and thus available from `DAO`. The only other thing which could be needed is `GetConfig()`.

**Comment by roman-khimov** [comment]
Let's postpone this a little, I need to go through interops again and check what's the best solution. We have some hierarchy in `Store`->`DAO`->`Blockchain` where each layer solves some part of the overall problem (`Store` is a generic KV store, `DAO` is an object layer around `Store` and `Blockchain` is a blockchain where you only change it by adding blocks/headers, but can get a lot of information about its state). At the same time there are some complications around that like `IsDoubleSpend` or `IsDoubleClaim` currently in `DAO` although they're higher level functions, there is caching at `Store` level and at the same time there is caching at `DAO` level.

So the question is whether interops should be using `Blockchain` in some form or `DAO`. I don't think they should use both as `Blockchain` in nothing without `DAO`, although at the moment we have it being used as a singleton with the one and only true `DAO` inside. Maybe that's not how `Blockchain` should be used. At the same time it shouldn't ever expose its `DAO` to the outside.

**Comment by roman-khimov** [comment]
OK, so interops really are special and can't work with just `Blockchain`. They change state directly via `PutStorageItem/PutContractState/PutAssetState` and can also delete some state information. One thing I've also noticed is that our current `GetValidators` implementation uses `dao`, but contract probably expects the (filtered) result from `Blockchain`. That probably suggests that we should avoid passing `Blockchain` there and just use `dao` interface plus some additional data  (cache, actually, because it can also be derived from `dao`) like current top block.

**Comment by roman-khimov** [comment]
OK, let's proceed with further improvements in some other tasks and merge this one as is (well, almost). It does solve the problem of interops being tied to the `core` and that's a huge win already.

File: pkg/core/interops.go [hunk 1]
```diff
@@ -66,213 +66,215 @@ type interopedFunction struct {
...
-		return &vm.InteropFuncPrice{Func: func(v *vm.VM) error {
-			return slice[n].Func(ic, v)
-		}, Price: slice[n].Price}
+func getInteropFromSlice(ic *interopContext, slice []interopedFunction) func(uint32) *vm.InteropFuncPrice {
+	return func(id uint32) *vm.InteropFuncPrice {
+		n := sort.Search(len(slice), func(i int) bool {
+			return slice[i].ID >= id
+		})
+		if n < len(slice) && slice[n].ID == id {
+			// Currying, yay!
```
Comment (roman-khimov) [inline pkg/core/interops.go:58]: Well, no more currying here after `bad39dd8863e32fbfa64d0e9fac9f1ec3448c678` (you see, that why it starts with `bad`!), so the comment is no longer relevant.

**Pull request merged**